# Bank Loan Default Risk Study

In [10]:
# импортируем библиотеку pandas
import pandas as pd

In [11]:
# прочитаем csv-файл
data = pd.read_csv('../datasets/bank_loaners.csv')

In [12]:
# Выведем первые 20 строчек датафрейма data на экран.
data.head(20)

Unnamed: 0  children  days_employed  dob_years            education  \
0            0         1   -8437.673028         42               высшее   
1            1         1   -4024.803754         36              среднее   
2            2         0   -5623.422610         33              Среднее   
3            3         3   -4124.747207         32              среднее   
4            4         0  340266.072047         53              среднее   
5            5         0    -926.185831         27               высшее   
6            6         0   -2879.202052         43               высшее   
7            7         0    -152.779569         50              СРЕДНЕЕ   
8            8         2   -6929.865299         35               ВЫСШЕЕ   
9            9         0   -2188.756445         41              среднее   
10          10         2   -4171.483647         36               высшее   
11          11         0    -792.701887         40              среднее   
12          12         0            NaN         65              среднее   
13          13         0   -1846.641941         54  неоконченное высшее   
14          14         0   -1844.956182         56               высшее   
15          15         1    -972.364419         26              среднее   
16          16         0   -1719.934226         35              среднее   
17          17         0   -2369.999720         33               высшее   
18          18         0  400281.136913         53              среднее   
19          19         0  -10038.818549         48              СРЕДНЕЕ   

    education_id     family_status  family_status_id gender income_type  debt  \
0              0   женат / замужем                 0      F   сотрудник     0   
1              1   женат / замужем                 0      F   сотрудник     0   
2              1   женат / замужем                 0      M   сотрудник     0   
3              1   женат / замужем                 0      M   сотрудник     0   
4              1  гражданский брак                 1      F   пенсионер     0   
5              0  гражданский брак                 1      M   компаньон     0   
6              0   женат / замужем                 0      F   компаньон     0   
7              1   женат / замужем                 0      M   сотрудник     0   
8              0  гражданский брак                 1      F   сотрудник     0   
9              1   женат / замужем                 0      M   сотрудник     0   
10             0   женат / замужем                 0      M   компаньон     0   
11             1   женат / замужем                 0      F   сотрудник     0   
12             1  гражданский брак                 1      M   пенсионер     0   
13             2   женат / замужем                 0      F   сотрудник     0   
14             0  гражданский брак                 1      F   компаньон     1   
15             1   женат / замужем                 0      F   сотрудник     0   
16             1   женат / замужем                 0      F   сотрудник     0   
17             0  гражданский брак                 1      M   сотрудник     0   
18             1    вдовец / вдова                 2      F   пенсионер     0   
19             1         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   7

In [13]:
# Выведем основную информацию о датафрейме с помощью метода info()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        21525 non-null  int64  
 1   children          21525 non-null  int64  
 2   days_employed     19351 non-null  float64
 3   dob_years         21525 non-null  int64  
 4   education         21525 non-null  object 
 5   education_id      21525 non-null  int64  
 6   family_status     21525 non-null  object 
 7   family_status_id  21525 non-null  int64  
 8   gender            21525 non-null  object 
 9   income_type       21525 non-null  object 
 10  debt              21525 non-null  int64  
 11  total_income      19351 non-null  float64
 12  purpose           21525 non-null  object 
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB


## Предобработка данных

### Удаление пропусков

In [14]:
# Выведем количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.
data.isna().sum()

Unnamed: 0             0
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [15]:
# 1. Выыедем список уникальных значений столбца Income_type для хардкода или для анализа
data['income_type'].unique()
for income_type in data['income_type'].unique():
    data.loc[(data['income_type'] == income_type) & (data['total_income'].isna()), 'total_income'] =  data.loc[(data['income_type'] == income_type), 'total_income'].median()     
        

### Обработка аномальных значений

In [16]:
# В данных могут встречаться артефакты (аномалии), например отрицательное количество дней трудового стажа 
# в столбце days_employed. Обработаем значения в этом столбце: 
# замените все отрицательные значения положительными с помощью метода abs().

data['days_employed'] = data['days_employed'].abs()

In [17]:
# Для каждого типа занятости выведите медианное значение трудового стажа days_employed в днях.

    
print(data.groupby('income_type')['days_employed'].median())

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64


У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

In [18]:
# Выведем перечень уникальных значений столбца children.

print(data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


In [19]:
# В столбце children есть два аномальных значения. 
# Удалим такие строки
data = data[(data['children'] >= 0) & (data['children'] < 20 )] 



In [20]:
# Выведем перечень уникальных значений столбца children, чтобы убедиться, что артефакты удалены.
print(data['children'].unique())

[1 0 3 2 4 5]


In [21]:
# Заполним пропуски в столбце days_employed медианными значениями по каждому типу занятости income_type.
for income_type in data['income_type'].unique():
        data.loc[(data['income_type'] == income_type) & (data['days_employed'].isna()), 'days_employed'] =  data.loc[(data['income_type'] == income_type), 'days_employed'].median()
        
  

In [22]:
# Убедимся, что все пропуски заполнены. выведем количество пропущенных значений для каждого столбца 
data.isna().sum()

Unnamed: 0          0
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [23]:
# Заменим вещественный тип данных в столбце total_income на целочисленный
data['total_income'] = data['total_income'].astype('int')

### Обработка дубликатов

In [24]:
# Обработаем неявные дубликаты в столбце education. 
#В этом столбце есть одни и те же значения, но записанные по-разному: 
#с использованием заглавных и строчных букв. 

data['education'] = data['education'].str.lower()

In [25]:
#  Выведем  количество строк-дубликатов в данных. 
data['education'].duplicated().sum()

21397

In [26]:
# удалим дубликаты

data = data.drop_duplicates()

### Категоризация данных

**Добавим столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [27]:
def categorize_income(total_income):

# Ф-ия возвращает кредитную категорию исходя из критерия сумарного дохода:
#    доход в диапазоне 0–30000 — 'E';
#    доход в диапазоне 30001–50000 — 'D';
#    доход в диапазоне 50001–200000 — 'C';
#    доход в диапазоне 200001–1000000 — 'B';
#    доход в диапазоне 1000001 и выше — 'A'.

    if (total_income >= 0) & (total_income <= 30000):
        return 'E'

    if (total_income >= 30001) & (total_income <= 50000):
        return 'D'
    
    if (total_income >= 50001) & (total_income <= 200000):
        return 'C'
    
    if (total_income >= 200001) & (total_income <= 1000000):
        return 'B'
    
    if (total_income >= 1000001):
        return 'A'


In [28]:
# применим функцию методом apply()
data['total_income_category'] = data['total_income'].astype('int').apply(categorize_income)

In [29]:
# Выведем на экран перечень уникальных целей взятия кредита из столбца purpose
print(data['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


**Создам функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.


In [30]:
# создайте функцию categorize_purpose()
def categorize_purpose(purpose):
    if 'автом' in purpose:
        return 'операции с автомобилем'
    if ('образо' in purpose) | ('учеб' in purpose ):
        return 'получение образования'
    if ('свадь' in purpose) | ('брако' in purpose ):
        return 'проведение свадьбы'
    if ('жиль' in purpose ) | ('едвиж' in purpose ) :
        return 'операции с недвижимостью'

In [31]:
# примените функцию методом apply()
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [37]:

debt_from_children = pd.DataFrame()
debt_from_children['debt_sum'] = data.groupby('children')['debt'].sum()
debt_from_children['category_count'] = data.groupby('children')['debt'].count()
debt_from_children['result_children'] = debt_from_children['debt_sum'] / debt_from_children['category_count'] 
debt_from_children.sort_values('result_children', ascending = False)


debt_sum  category_count  result_children
children                                           
4                4              41         0.097561
2              194            2055         0.094404
1              444            4818         0.092154
3               27             330         0.081818
0             1063           14149         0.075129
5                0               9         0.000000

**Промежуточный вывод**

Если сравнить только самые многочисленные выборки с 0, 1 и 2 детьми, можно сказать, что чем больше у заемщика детей, тем выше риск допустить просрочку

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [38]:
debt_from_family_status = pd.DataFrame()
debt_from_family_status['sum_family_status'] =data.groupby('family_status')['debt'].sum()
debt_from_family_status['count_family_status'] = data.groupby('family_status')['debt'].count()
debt_from_family_status['result_family_status'] = debt_from_family_status['sum_family_status'] / debt_from_family_status['count_family_status'] 
debt_from_family_status.sort_values('result_family_status', ascending = False)


sum_family_status  count_family_status  \
family_status                                                   
Не женат / не замужем                273                 2799   
гражданский брак                     385                 4160   
женат / замужем                      927                12302   
в разводе                             84                 1189   
вдовец / вдова                        63                  952   

                       result_family_status  
family_status                                
Не женат / не замужем              0.097535  
гражданский брак                   0.092548  
женат / замужем                    0.075354  
в разводе                          0.070648  
вдовец / вдова                     0.066176

**Промежуточный вывод**

Люди не в браке и не бывавшие в браке имеют больший процент невозвратов в срок. Но, те кто развелись или овдовели чаще платят в срок, чем люди в браке. 

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [40]:
debt_from_total_income = pd.DataFrame()

#debt_from_total_income['cat'] = data.groupby('total_income_category')['debt']
debt_from_total_income['debt_sum'] = data.groupby('total_income_category')['debt'].sum()
debt_from_total_income['category_count'] = data.groupby('total_income_category')['debt'].count()
debt_from_total_income['conversion'] = (debt_from_total_income['debt_sum'] / debt_from_total_income['category_count']) * 100;
debt_from_total_income['conversion'] = debt_from_total_income['conversion'].map('{:,.2f} %'.format)
debt_from_total_income.sort_values('total_income_category', ascending = True)

print(debt_from_total_income) 
print('-----------------------------')
print("A Доход свыше: 1'000'001.00")
print("B Доход от 200'001.00 до 1'000'000.00")
print("C Доход от: 50'001.00 до 200'000.00")
print("D Доход от 30'001.00 до 200'000.00")
print("E Доход до 30'000.00")      


                       debt_sum  category_count conversion
total_income_category                                     
A                             2              25     8.00 %
B                           354            5014     7.06 %
C                          1353           15992     8.46 %
D                            21             349     6.02 %
E                             2              22     9.09 %
-----------------------------
A Доход свыше: 1'000'001.00
B Доход от 200'001.00 до 1'000'000.00
C Доход от: 50'001.00 до 200'000.00
D Доход от 30'001.00 до 200'000.00
E Доход до 30'000.00


**Промежуточный вывод**

Если сравнить только две самые большие категории C (50 - 200) и B (200 - 1 млн), выходит что заемщики с более высоким доходом являются более надежными.

Также в качестве рекомендации можно предложить банку изменить градацию дохода, чтобы получить более сбалансированные выборки.

### Как разные цели кредита влияют на его возврат в срок?

In [41]:
debt_from_purpose_category = pd.DataFrame()
debt_from_purpose_category['sum_purpose_category'] = data.groupby('purpose_category')['debt'].sum()
debt_from_purpose_category['count_purpose_category'] = data.groupby('purpose_category')['debt'].count()
debt_from_purpose_category['result_purpose_category'] = debt_from_purpose_category['sum_purpose_category'] / debt_from_purpose_category['count_purpose_category'] 
debt_from_purpose_category.sort_values('result_purpose_category', ascending = False)


sum_purpose_category  count_purpose_category  \
purpose_category                                                         
операции с автомобилем                     400                    4288   
получение образования                      369                    3997   
проведение свадьбы                         183                    2337   
операции с недвижимостью                   780                   10780   

                          result_purpose_category  
purpose_category                                   
операции с автомобилем                   0.093284  
получение образования                    0.092319  
проведение свадьбы                       0.078306  
операции с недвижимостью                 0.072356

**Промежуточный вывод**

Вывод: самые ненадежные заемщики те, кто берут кредиты на образование (доля просрочки 0.092) и покупку автомобиля (0.93); далее по росту надежности следуют заемщики, которые берут кредиты на проведение свадьбы (доля просрочки = 0.079); самые надежные заемщики - те, кто берут кредиты на операции с недвижимостью (доля просрочки = 0.072).


### Проанализируем возможные причины появления пропусков в исходных данных

Причины могут быть следующими:

Ошибка выгрузки, проблеммы при передачи, ошибки форматов, разные локализации систем, человеческий фактор данные никогда не вводились оператором


### Попытаемся объяснить почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства, поэтому медианной можно вычислить наиболее достовреное значение, особенно в случаях когда есть возможность сужать поиск медианы по определенным категориям


# Заключение

Целью нашего исследования было изучение платежеспособности различных груп населения в зависимости от видов кредита для того чтобы выявить тренды и дать рекомендации по минимизации рисков кредитным учреждениям. 

- наиболее привлекательными для кредитных учреждений, являются ипотечные кредиты семьям без детей. Здесь наблюдаются наименьшие риски связанные с выплатам в срок.
- семейный статус влияет на вероятность платежей по кредиту в срок - люди, бывшие в официальном браке чаще плятят в срок чем те, кто не бывали в нем, причем разведенные и овдовевшие платят в срок чаще, чем люди в браке.
- количество детей влияет на платежеспособность. Чем больше детей, тем выше риск просрочки платежей. Семьи без детей реже просрачивают оплату.

**Рекомендации банку** 

Основываясь на проведенном анализе предоставленных данных наиболее перспективных заемщиков следует отбирать по следующим признакам:

- желающие взять ипотечный кредит (не автокредит, не кредит на образование);
- семейные люди без детей;
